### ISSUE TITLE
DP with *_step_end incorrect reduce

### ISSUE THREAD

**Issue by chris-tkinter** [PyTorchLightning/pytorch-lightning#10634]
## 🐛 Bug

From the documentation the input of `*_step_end` is a collection of sub-batches. Thus for example, if the `*_step` return loss and the model is trained on 4 gpus, then `*_step_end` should return a loss tensor of shape `(4, )`, 1 for each gpu.

However, when I am doing
```

    def validation_step(self, batch: Any, batch_idx: int):
    
        loss, pred = self(batch)
        return {"loss": loss, "pred": pred, "label": batch["label"]}

    def validation_step_end(self, output: dict) -> dict:
        ### HERE
        return output
```
the input of `validation_step_end` has incorrect shape. I am training on 4 gpus. On my end, the loss is `(4, )` tensor which is correct. But `output['pred']` only has 2 tensors which is not correct (it should have 4); whereas `output['label']` has 8 tensors. I am wondering what might cause this?

### Expected behavior
`output['pred']` should has 4 tensors (1 from each gpu) and  `output['label']`  should has 4 tensors (1 from each gpu)
 
### Environment

- PyTorch Lightning Version (e.g., 1.3.0): 1.5.2
- PyTorch Version (e.g., 1.8) 1.9.1
- Python version: 3.7.11

cc @justusschock @awaelchli @tchaton

**Comment by awaelchli** [comment]
Hi @chris-tkinter
Can you provide an example based on [this code here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py) with dummy values for the data you pass through these hooks? 
You didn't mention it, but this issue applies for `accelerator="dp" / strategy="dp"`, right? 

Note that anything returned in the step methods gets processed through the `torch.nn.DataParallel` and then its output gets passed to `training_step_end`.

**Comment by awaelchli** [comment]
Hi again, I didn't get a reply from you yet. I'm not able to reproduce what you describe with just the out of context code snippet. 

Note that if the dataset size is not divisible by the batch size, it is possible that the last batch is smaller than the number of GPUs (4  in your case). This means the last batch will run on less than 4 GPUs and thus you will get less than 4 predictions. Is it possible you ran into this edge case? You can set` drop_last=True` in the DataLoader.
